From 5bf43afd8d571947d0880a20001938b3cf750bc8 Mon Sep 17 00:00:00 2001
From: Al Viro <viro@zeniv.linux.org.uk>
Date: Mon, 4 Jan 2021 15:25:34 -0500
Subject: [PATCH 3/4] BACKPORT: umount(2): move the flag validity checks first

Unfortunately, there's userland code that used to rely upon these
checks being done before anything else to check for UMOUNT_NOFOLLOW
support.  That broke in c762edd9b055 ("BACKPORT: fs: refactor ksys_umount").
Separate those from the rest of checks and move them to ksys_umount();
unlike everything else in there, this can be sanely done there.

Reported-by: Sargun Dhillon <sargun@sargun.me>
Fixes: c762edd9b055 ("BACKPORT: fs: refactor ksys_umount")
Change-Id: I7c4487567779ca7cc1e6d41c2623610547efbcd1
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Signed-off-by: mrsrimar22 <mrsrimar22@gmail.com>
---
 fs/namespace.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/fs/namespace.c b/fs/namespace.c
index 4b9b15653cc4..0d3267c06e63 100644
--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -1700,8 +1700,6 @@ static int can_umount(const struct path *path, int flags)
 {
 	struct mount *mnt = real_mount(path->mnt);
 
-	if (flags & ~(MNT_FORCE | MNT_DETACH | MNT_EXPIRE | UMOUNT_NOFOLLOW))
-		return -EINVAL;
 	if (!may_mount())
 		return -EPERM;
 	if (path->dentry != path->mnt->mnt_root)
@@ -1716,6 +1714,7 @@ static int can_umount(const struct path *path, int flags)
 	return 0;
 }
 
+// caller is responsible for flags being sane
 int path_umount(struct path *path, int flags)
 {
 	struct mount *mnt = real_mount(path->mnt);
@@ -1738,6 +1737,10 @@ int ksys_umount(char __user *name, int flags)
 	int ret;
 	int lookup_flags = 0;
 
+	// basic validity checks done first
+	if (flags & ~(MNT_FORCE | MNT_DETACH | MNT_EXPIRE | UMOUNT_NOFOLLOW))
+		return -EINVAL;
+
 	if (!(flags & UMOUNT_NOFOLLOW))
 		lookup_flags |= LOOKUP_FOLLOW;
 
-- 
2.52.0

